### ISSUE TITLE
Add syntax highlighting support for languages inside R strings

### ISSUE THREAD

**Issue by talgalili** [rstudio/rstudio#2232]
When using (for example) DBI with RSQLite, I tend to write SQL code inside a string, and then supply it to `dbGetQuery`. Obviously, RStudio already has syntax highlighting for SQL, but it is only applied when used inside a dedicated code chunk (i.e.: https://rmarkdown.rstudio.com/authoring_knitr_engines.html#sql).

What I would like is that in the following code chunk, the SQL string would be automatically highlighted (and formatted?) making it simpler for me to edit it inside RStudio:

```r

library(DBI)
mydb <- dbConnect(RSQLite::SQLite(), "")
dbWriteTable(mydb, "mtcars", mtcars)

# lang: SQL
sql_query <- " --SQL
  SELECT * 
  FROM mtcars 
  WHERE vs=0
  LIMIT 5
  "

dbGetQuery(mydb, sql_query)

```

The one thing I'm not sure is what would be the best way to tell RStudio it should use SQL syntax highlighting. One way is to have a comment before the string with something like #lang: SQL, or alternatively, it could be that the string itself should have something like --SQL. Alternatively, since SQL always has SELECT and FROM, maybe to just detect strings that include these keywords and operate on them.

Either way, adding this would be very helpful to me in my work. Thanks :)

**Comment by kevinushey** [comment]
Thanks for the feature request. Unfortunately, this would be quite expensive to implement and we feel that R Markdown documents are a better mechanism for hosting multi-language documents.

**Comment by talgalili** [comment]
Hi @kevinushey - thanks for the quick response and explanation.
Alternatively, would it be possible (if it is not already) to add a chunk parameter to control the syntax highlighting of that chunk?
For example, if I use the following in an Rmd file, it would have an SQL syntax hilighting:
```
```{SQL}
SELECT *
FROM tbl
```

```

What I would like is to have something like:

```
```{r syntax_lang = "SQL"}
txt_Sql <- "
SELECT *
FROM tbl
"
```

```

This way, it would run the chunk as an R code, but while I'm editing the SQL query, it would be easier for me to look at the code.

What do you think?

**Comment by talgalili** [comment]
Hi @kevinushey - thanks for the quick response and explanation.
Yesterday I've had a lovely conversation with several RStudio folks, and I hope it might be possible to get something like this to work (assuming some token in the string could be used to help RStudio detect it is an SQL query. For example --SQL or something similar)

**Comment by jmcphers** [comment]
Where does the `--SQL` convention come from here?

**Comment by talgalili** [comment]
If you're asking about the sql flavour I get to use - it is presto or hive
usually.
If you ask about the tag for identification of the text as SQL, I didn't
rely on a convention but just tried to convey the idea simply. If there is
an existing standard for this than obviously it would be better to use it.

**Comment by jmcphers** [comment]
> If there is an existing standard for this than obviously it would be better to use it.

Yes, that's what I'm getting at! I haven't been able to find one, however. Have you ever used an IDE which has the behavior you're seeking?

**Comment by gtritchie** [comment]
Closest thing I’ve seen are the language injection comments and annotations in Jetbrains IDE’s. For some details: https://www.jetbrains.com/help/idea/using-language-injections.html

**Comment by talgalili** [comment]
Hi @jmcphers 
I see this in atom IDE (mac), when editing a python script with a string starting as (say)
x = """
SELECT
 * 
FROM 
 tbl
"""
It will know to detect it is SQL. If I remove the x = , it will no longer highlight the SQL, so I assume that this combination is what is used to know which highlighting to use.
However, I don't know which atom package is responsible for this behavior.

**Comment by dwy904** [comment]
any update from this?

**Comment by rld2** [comment]
I'm very interested in something like this. I primarily work in Rmarkdown. For me, a better solution would be able to use the sql chunk (which has highlighting) but add the option to save the sql string instead of the sql result. I could manipulate the resulting string with much more flexibility and pass to a custom sql retrieving function.

**Comment by yihui** [comment]
@rld2 You can use chunk references (<https://yihui.name/knitr/demo/reference/>), e.g.,

````md
```{sql, chunk-a, eval=FALSE}
select * from tbl
```

```{r, chunk-b}
x = "
<<chunk-a>>
"
```
````

**Comment by boshek** [comment]
+1 for this feature. @talgalili idea is an attractive one and could possibly leverage improvements merged here: https://github.com/rstudio/rstudio/pull/3429 (<- thanks!)

From a workflow perspective, I don't think all uses of R and SQL can fall into the Rmarkdown workflow so I think this need will remain and result in folks employing a hodge podge of solutions. Having this implemented would remove this need.

**Comment by moodymudskipper** [comment]
Answering @kevinushey 's comment :

> Thanks for the feature request. Unfortunately, this would be quite expensive to implement and we feel that R Markdown documents are a better mechanism for hosting multi-language documents.

This might be true for python but personally I have SQL code all over my projects, functions, shiny apps... Markdown is not always the answer.

Another situation is writing Rcpp functions, with current highlighting it seems really wrong : 

```
cppFunction('int signC(int x) {
  if (x > 0) {
    return 1;
  } else if (x == 0) {
    return 0;
  } else {
    return -1;
  }
}')
```

The highlighting would be a great help (and spotting invalid syntax would be an amazing improvement too).

Yet another similar situation is writing java script and html for shiny apps. We can also sometime have R code in strings.

For what it's worth, I like the approach of @talgalili and I think we can copy what Jetbrain does on @gtritchie 's link

```
# language=SQL
sql_query <- "
SELECT * 
FROM mtcars 
WHERE vs=0
LIMIT 5
"
```

Ideally : 

* The string literals found in the command following the comment are subject to the special highlighting
* We also get special indentation if relevant for given language
* We also get syntax checking (by icon in margin, or background coloring etc)

## A special case for SQL ?

I believe SQL is the language that is mostly used in combination with R.

In Options/Code/Display We could have an additional checkbox, unchecked by default, labelled "Highlight SQL query strings".

If it is checked, literal strings in a script will be tested and if they are valid SQL, surrounding quotes will be greyed out and the SQL will be formatted as it is in SQL chunks of Markdown reports.

The recognizable format of SQL command makes it unlikely to often encounter false positives (e.g. a string "select item from menu", not meant to be sql).

**Comment by henningsway** [comment]
I default to RMarkdown for most of my interactive development. When I use {sql}-chunks I appreciate the syntax highlighting. But I can only run them, when I set `chunk_output_type:` to ` inline` instead of `console`.

From my point of view being able to use syntax-highlighted sql independently of "inline-output-notebooks" seems useful. :)

**Comment by pnacht** [comment]
I've already +1'd this feature request but would also like to offer another solution.

Per @kevinushey, requiring RStudio to parse SQL statements for validity would get expensive. I'm personally forced to work with humongous code files with many thousands of lines, much of which is embedded SQL. So I can see how parsing that for validity would get expensive (though the debug advantage mentioned by @moodymudskipper would be nice).

One solution would be for the R team to add a shorthand, as they did with `r"()"` for raw strings in 4.0.0. Maybe `s"()"`. But obviously this isn't the proper forum to talk about changes to the R language itself.

So a temporary solution would be for RStudio to format any string found within specific function calls, such as DBI::SQL() and DBI::dbGetQuery(), or any function which will basically only ever be called with SQL strings. I'm no expert 

If this is an appealing idea, I'd also ask that the list of functions which get formatted be extendable (even if via a config file), since I personally work with multiple wrapper functions around DBI calls (I can't remember the last time I called `DBI::dbGetQuery()` directly).

**Comment by colearendt** [comment]
Love this idea
